app/models/concerns/digital_object_concerns/digital_object_data/setters/projects.rb

```diff
@@ -6,8 +6,18 @@ module Projects
       extend ActiveSupport::Concern
 
       def set_projects(digital_object_data)
-        return unless digital_object_data.key?('projects')
-        self.projects = Set.new(digital_object_data['projects'].map { |digital_object_data_project|
+        set_primary_project(digital_object_data)
+        set_other_projects(digital_object_data)
+      end
+
+      def set_primary_project(digital_object_data)
+        return unless digital_object_data.key?('primary_project')
+        self.primary_project = Project.find_by(string_key: digital_object_data['primary_project']['string_key'])
+      end
+
+      def set_other_projects(digital_object_data)
+        return unless digital_object_data.key?('other_projects')
+        self.other_projects = Set.new(digital_object_data['other_projects'].map { |digital_object_data_project|
```

Comment:
In the case of a project string key typo, this would result in nil elements being saved in the other_projects array (unless there's another protection elsewhere in the code that I'm not thinking of / haven't seen yet). We should raise an exception or have a validation for this so we can somehow inform the user of their error. I think that an exception might be easier to handle, since a validation could only tell the user that one of their other_projects is nil, and not which original invalid string value resulted in a conversion to nil.

Reply:
yes, the validation would be too late unless we actually stored a jsonified "bad project" entry. I'll look into an exception.